android-components/components/feature/app-links/src/test/java/mozilla/components/feature/app/links/AppLinksFeatureTest.kt

```diff
@@ -222,6 +223,135 @@ class AppLinksFeatureTest {
         verify(mockDialog, never()).showNow(eq(mockFragmentManager), anyString())
     }
 
+    @Test
+    fun `WHEN user rejected external app THEN an external app dialog is not shown`() {
+        feature = spy(
+            AppLinksFeature(
+                context = mockContext,
+                store = store,
+                fragmentManager = mockFragmentManager,
+                useCases = mockUseCases,
+                dialog = mockDialog,
+                loadUrlUseCase = mockLoadUrlUseCase,
+                shouldPrompt = { true },
+            ),
+        ).also {
+            it.start()
+        }
+
+        val tab = createTab(webUrl)
+        val appIntent: Intent = mock()
+        val componentName: ComponentName = mock()
+        doReturn(componentName).`when`(appIntent).component
+        doReturn("com.zxing.app").`when`(componentName).packageName
+        AppLinksFeature.doNotOpenCacheMap[("com.zxing.app" + tab.id).hashCode()] = System.currentTimeMillis()
+
+        feature.handleAppIntent(tab, intentUrl, appIntent)
+        verify(mockDialog, never()).showNow(eq(mockFragmentManager), anyString())
+    }
+
+    @Test
+    fun `WHEN user rejected external app on one tab THEN external app dialog will still show on another tab`() {
```

Comment:
totally out of scope, but it could be nice to extend this behavior such that tabs for search engines didn't show the dialog.

for example, I might be clicking in reddit links in google searches across various tabs, but don't want to get the dialog every time on new search tabs.

Reply:
Good idea.  But this will need some thinking.  A lot of redirect happens before reaching a site, such as Amazon, Instagram etc.  User is prompted before reaching the site so they start they experience directly in the official app.   In this change I added tab ID as part of the key because I wasn't sure if a decision made in one tab should apply to other similar tabs.